### ISSUE TITLE
 s3utils.IsValidDomain rejects fully qualified domain names

### ISSUE THREAD

**Issue by teejaded** [minio/minio-go#1109]
What happened:

s3utils.IsValidDomain returns false for FQDNs. This is a domain name that ends in a dot -- signifying that it is not relative to a DNS suffix search list.

What I expected:

s3utils.IsValidDomain should accept FQDNs

https://github.com/minio/minio-go/blob/master/pkg/s3utils/utils.go#L50

**Comment by harshavardhana** [comment]
Please a sample program on what are you trying to do @teejaded ?

**Comment by teejaded** [comment]
Sure.

```golang
package main

import (
	"fmt"

	"github.com/minio/minio-go"
)

func main() {
	_, err := minio.New("rook-ceph-obj.rook-ceph.svc.cluster.local.", "MYKEY", "MYKEY", true)
	if err != nil {
		fmt.Printf("Error creating s3 client: %s", err)
	}
}
```

Output:

```
./minio-test
Error creating s3 client: Endpoint: rook-ceph-obj.rook-ceph.svc.cluster.local. does not follow ip address or domain name standards.
```

**Comment by harshavardhana** [comment]
> ```go
> 	_, err := minio.New("rook-ceph-obj.rook-ceph.svc.cluster.local.", "MYKEY", "MYKEY", true)
> ```

Having a `.` at the end is not a valid domain @teejaded so it's correctly complaining.

**Comment by teejaded** [comment]
I disagree.  For example, this works as I would expect:  https://google.com./

The trailing dot indicates that the name starts at the root domain.  Using it prevents DNS clients in Kubernetes from making several extra dns queries.

http://www.dns-sd.org/trailingdotsindomainnames.html

**Comment by harshavardhana** [comment]
> I disagree. For example, this works as I would expect: https://google.com./
> 

Let's not go to DNS compatibilities yet. If a `.` can work with the root domain, so should be `google.com` so are you saying that removing `.` is not an option?

> The trailing dot indicates that the name starts at the root domain. Using it prevents DNS clients in Kubernetes from making several extra dns queries.
> 

Can you explain what benefit does it add in more detail?

**Comment by teejaded** [comment]
Here some more detail into why this matters to me.  First lets look at the generated `/etc/resolv.conf` inside a pod running in a Kubernetes cluster.

```
nameserver 10.96.0.10
search default.svc.cluster.local svc.cluster.local cluster.local
options ndots:5
```

There's an interesting setting here `options ndots:5`.  Here's the manpage entry for this option.

```text
  ndots:n
         Sets a threshold for the number of dots which must appear in a name given to
         res_query(3) (see resolver(3)) before an initial absolute query will be made.  The
         default for n is 1, meaning that if there are any dots in a name, the name will be
         tried first as an absolute name before any search list elements are appended to it.
         The value for this option is silently capped to 15.
```



Here's the behavior in action.

```
$ kubectl run --rm -it --image ubuntu:latest tj-test -- /bin/bash
kubectl run --generator=deployment/apps.v1 is DEPRECATED and will be removed in a future version. Use kubectl run --generator=run-pod/v1 or kubectl create instead.
If you don't see a command prompt, try pressing enter.


root@tj-test-56f8cfd77c-nmrhp:/# apt-get update && apt-get install -y dnsutils tcpdump

...snip...

root@tj-test-56f8cfd77c-nmrhp:/# tcpdump -vvv udp port 53 &
[1] 472
root@tj-test-56f8cfd77c-nmrhp:/# tcpdump: listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes

root@tj-test-56f8cfd77c-nmrhp:/# nslookup google.com > /dev/null
16:59:13.558515 IP (tos 0x0, ttl 64, id 55070, offset 0, flags [none], proto UDP (17), length 82)
    tj-test-56f8cfd77c-nmrhp.49182 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93cb -> 0xe0cc!] 12613+ A? google.com.default.svc.cluster.local. (54)
16:59:13.558996 IP (tos 0x0, ttl 63, id 31535, offset 0, flags [DF], proto UDP (17), length 175)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.49182: [bad udp cksum 0x9428 -> 0x47f2!] 12613 NXDomain*- q: A? google.com.default.svc.cluster.local. 0/1/0 ns: cluster.local. [5s] SOA ns.dns.cluster.local. hostmaster.cluster.local. 1558566340 7200 1800 86400 5 (147)
16:59:13.559279 IP (tos 0x0, ttl 64, id 55071, offset 0, flags [none], proto UDP (17), length 74)
    tj-test-56f8cfd77c-nmrhp.32806 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93c3 -> 0x0a37!] 64541+ A? google.com.svc.cluster.local. (46)
16:59:13.559649 IP (tos 0x0, ttl 64, id 55072, offset 0, flags [DF], proto UDP (17), length 69)
    tj-test-56f8cfd77c-nmrhp.54456 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93be -> 0xe8f4!] 37757+ PTR? 10.0.96.10.in-addr.arpa. (41)
16:59:13.559807 IP (tos 0x0, ttl 62, id 65012, offset 0, flags [DF], proto UDP (17), length 167)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.32806: [udp sum ok] 64541 NXDomain*- q: A? google.com.svc.cluster.local. 0/1/0 ns: cluster.local. [5s] SOA ns.dns.cluster.local. hostmaster.cluster.local. 1558566340 7200 1800 86400 5 (139)
16:59:13.559989 IP (tos 0x0, ttl 64, id 55073, offset 0, flags [none], proto UDP (17), length 70)
    tj-test-56f8cfd77c-nmrhp.41504 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93bf -> 0x18d4!] 41480+ A? google.com.cluster.local. (42)
16:59:13.560224 IP (tos 0x0, ttl 63, id 31536, offset 0, flags [DF], proto UDP (17), length 163)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.41504: [bad udp cksum 0x941c -> 0x7ff9!] 41480 NXDomain*- q: A? google.com.cluster.local. 0/1/0 ns: cluster.local. [5s] SOA ns.dns.cluster.local. hostmaster.cluster.local. 1558566340 7200 1800 86400 5 (135)
16:59:13.561133 IP (tos 0x0, ttl 62, id 65013, offset 0, flags [DF], proto UDP (17), length 82)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.41520: [udp sum ok] 4481 q: A? google.com. 1/0/0 google.com. [30s] A 216.58.192.14 (54)
16:59:13.561586 IP (tos 0x0, ttl 64, id 55075, offset 0, flags [none], proto UDP (17), length 56)
    tj-test-56f8cfd77c-nmrhp.40112 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93b1 -> 0x63ab!] 22469+ AAAA? google.com. (28)
16:59:13.584110 IP (tos 0x0, ttl 62, id 13076, offset 0, flags [DF], proto UDP (17), length 94)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.40112: [udp sum ok] 22469 q: AAAA? google.com. 1/0/0 google.com. [30s] AAAA 2607:f8b0:4005:801::200e (66)
root@tj-test-56f8cfd77c-nmrhp:/# nslookup google.com. > /dev/null
16:59:19.421809 IP (tos 0x0, ttl 64, id 56523, offset 0, flags [none], proto UDP (17), length 56)
    tj-test-56f8cfd77c-nmrhp.33921 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93b1 -> 0x2bf9!] 42945+ A? google.com. (28)
16:59:19.422820 IP (tos 0x0, ttl 63, id 32712, offset 0, flags [DF], proto UDP (17), length 82)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.33921: [bad udp cksum 0x93cb -> 0x0020!] 42945 q: A? google.com. 1/0/0 google.com. [30s] A 216.58.192.14 (54)
16:59:19.423175 IP (tos 0x0, ttl 64, id 56524, offset 0, flags [none], proto UDP (17), length 56)
    tj-test-56f8cfd77c-nmrhp.34480 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93b1 -> 0x85f2!] 19326+ AAAA? google.com. (28)
16:59:19.424375 IP (tos 0x0, ttl 62, id 65530, offset 0, flags [DF], proto UDP (17), length 94)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.34480: [udp sum ok] 19326 q: AAAA? google.com. 1/0/0 google.com. [30s] AAAA 2607:f8b0:4005:801::200e (66)
```

I count 6 A record requests before `google.com.` is found.  Now lets try with the trailing dot.

```
root@tj-test-56f8cfd77c-nmrhp:/# nslookup google.com. > /dev/null
16:59:32.924848 IP (tos 0x0, ttl 64, id 59318, offset 0, flags [none], proto UDP (17), length 56)
    tj-test-56f8cfd77c-nmrhp.35713 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93b1 -> 0x9e90!] 11818+ A? google.com. (28)
16:59:32.925388 IP (tos 0x0, ttl 62, id 16181, offset 0, flags [DF], proto UDP (17), length 82)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.35713: [udp sum ok] 11818 q: A? google.com. 1/0/0 google.com. [23s] A 216.58.192.14 (54)
16:59:32.925695 IP (tos 0x0, ttl 64, id 59319, offset 0, flags [none], proto UDP (17), length 56)
    tj-test-56f8cfd77c-nmrhp.48172 > kube-dns.kube-system.svc.cluster.local.53: [bad udp cksum 0x93b1 -> 0x6700!] 13556+ AAAA? google.com. (28)
16:59:32.925962 IP (tos 0x0, ttl 63, id 34372, offset 0, flags [DF], proto UDP (17), length 94)
    kube-dns.kube-system.svc.cluster.local.53 > tj-test-56f8cfd77c-nmrhp.48172: [bad udp cksum 0x93d7 -> 0x4c6c!] 13556 q: AAAA? google.com. 1/0/0 google.com. [23s] AAAA 2607:f8b0:4005:801::200e (66)
```

Just the 1 A request.

Honestly 6 DNS requests isn't the end of the world, but it is the beginning of a death by a thousand cuts as you begin to scale out.

**Comment by harshavardhana** [comment]
Understood @teejaded thanks for explaining I reproduced this on our k8s cluster - we are PR nonetheless with a fix #1110

**Comment by mqsoh** [comment]
This change doesn't help when the endpoint is an S3 address (and probably other providers). The [s3utils](https://github.com/minio/minio-go/blob/master/pkg/s3utils/utils.go) package has some regexes and string comparisons that break when the endpoint is `s3.amazonaws.com.` (trailing dot). I started working on a patch but got a little overwhelmed; I may contribute patch sometime later.

(Just like OP I'm running this in a Kubernetes cluster. Our staging cluster is hammering the VPCs nameserver with search queries and some queries are timing out. I'm updating our service URLs to use the FQDNs.)
